linkerd/client-policy/src/http/router.rs

```diff
@@ -0,0 +1,166 @@
+    }
+}
+
+// === impl ServiceRouter ===
+
+impl<T, N, S> ServiceRouter<T, N, S> {
+    fn update_route_policies(&mut self, route_policies: impl IntoIterator<Item = RoutePolicy>)
+    where
+        T: Clone,
+        N: NewService<(Option<RoutePolicy>, T), Service = S> + Clone,
+    {
+        for route in route_policies.into_iter() {
+            if let hash_map::Entry::Vacant(ent) = self.services.entry(route) {
+                let route = ent.key().clone();
+                let svc = self
+                    .new_route
+                    .new_service((Some(route), self.target.clone()));
+                ent.insert(svc);
+            }
+        }
+    }
+}
+
+impl<B, T, N, S> Service<http::Request<B>> for ServiceRouter<T, N, S>
+where
+    T: Clone,
+    N: NewService<(Option<RoutePolicy>, T), Service = S> + Clone,
+    S: Service<http::Request<B>> + Clone,
+    Error: From<S::Error>,
+{
+    type Response = S::Response;
+    type Error = Error;
+    type Future = MapErr<Oneshot<S, http::Request<B>>, fn(S::Error) -> Error>;
+
+    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Error>> {
+        let mut policy = match self.changed.poll_unpin(cx) {
+            Poll::Pending => return Poll::Ready(Ok(())),
+            Poll::Ready(update) => update?,
+        };
+
+        let http_routes = policy.borrow_and_update().http_routes.clone();
+
+        // If the routes have been updated, update the cache.
+        tracing::debug!(routes = %http_routes.len(), "Updating client policy HTTP routes");
+
+        // XXX(eliza): this will unify all routes that share the same policy to
+        // have a single service...which is nice, unless we want e.g. different
+        // metrics for each route. if we want that, we should probably include
+        // the metric labels in the policy, i think?
+        let route_policies = route_policies(&http_routes).collect::<HashSet<_>>();
+        self.http_routes = http_routes;
+
+        // Clear out defunct routes before building any missing routes.
+        self.services.retain(|r, _| route_policies.contains(r));
+        self.update_route_policies(route_policies);
+
+        // wait for the next change
+        self.changed.set(changed(policy));
+
+        Poll::Ready(Ok(()))
+    }
+
+    fn call(&mut self, req: http::Request<B>) -> Self::Future {
+        let inner = match super::find(&self.http_routes, &req) {
+            Some((req_match, policy)) => {
+                tracing::trace!(
+                    route.group = %policy.meta.group(),
+                    route.kind = %policy.meta.kind(),
+                    route.name = %policy.meta.name(),
+                    "req.match" = ?req_match,
+                    "Using HTTPRoute service",
+                );
+
+                self.services.get(policy).expect("route must exist").clone()
+            }
+            None => {
+                tracing::trace!("No HTTPRoutes matched");
+                self.default.clone()
```

Comment:
I'm pretty sure we have to error here in a way that causes a 404 from the error handler.

Reply:
ah, yup, we can change it to do that!

Reply:
this now 404s as of fc32de7a3ea7c57229e5f716730203ae7b93ef71